### PR TITLE
refactor(provider-bindgen): fix import & export semantics

### LINKED ISSUE #715: [BUG] Reverse the imports/exports WIT usage in bindgen
**Describe the bug**
At present the `provider-wit-bindgen` crate uses WIT imports and exports in a somewhat reversed manner. For example, the `kv-redis` provider world looks like the following:

```wit
package wasmcloud:provider-kvredis

world provider-kvredis {
    import wasmcloud:keyvalue/key-value
}
```

Given a WIT like the above, bindgen generates two important things:

- `InvocationHandler` structs methods to handle calling across the lattice (sending messages *out* on the lattice)
- `MessageDispatch#dispatch` implementation that listens for operations being sent across the lattice *to* the provider

Since components `export` functionality in order for other components to `import` that same functionality (and call/use it), the current behavior of `import`ing interfaces generating `MessageDispatch#dispatch` implementations is a awkward at best.

It would be better if `export`ing `wasmcloud:keyvalue/key-value` generated the dispatch handling, and `import`ing some other functionality generated the `InvocationHandler` functionality.

**To Reproduce**
1. See the WIT for providers that are bindgen'd

**Expected behavior**

`import`ed interfaces should generate `InvocationHandler`s, and `export`ed interfaces should generate incoming message handlers for lattice operations.

**Screenshots**

**Environment (please complete the following information) **
 - OS: [e.g. Linux, MacOS]
 - Shell [e.g. bash, zsh, powershell]
 - wasmcloud Version [e.g. 0.78.0]

**Additional context**

### PR DESCRIPTION
**Pull request by vados-cosmonic**
## Feature or Problem

The current semantics are unintuitive -- interfaces "exported" from a provider signifies functionality *provided* to the lattice, and the WIT worlds should reflect that.

At present, functionality exposed to the lattice for actors to use is signified by WIT *imports* in the provider worlds.

This commit reverses the import and export semantics that were used for WIT contracts in the providers, to make them more intuitive.

Resolves #715

### CODE REVIEW

**Review by brooksmtownsend — COMMENTED** [review]
This is really awesome stuff. I love the wits and some of the new config code, might warrant a second review from another maintainer more familiar with provider SDK but the semantics are great

**Review by brooksmtownsend — APPROVED** [review]
This is really awesome stuff. I love the wits and some of the new config code, might warrant a second review from another maintainer more familiar with provider SDK but the semantics are great

**Comment by connorsmith256** [comment]
@vados-cosmonic can you point to where the logical changes happened here, or split into multiple commits? With this being only one commit, I can't easily tell what was simply split into multiple files, and where the import/export semantics were reversed

File: crates/provider-wit-bindgen-macro/src/bindgen_visitor.rs [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+                                "adding exported trait method for path [{full_path}], trimmed signature: [{}]",
+                                trimmed_fn.sig.to_token_stream().to_string(),
+                            );
+                        self.export_trait_methods
+                            .entry(full_path)
+                            .or_default()
+                            .push(trimmed_fn);
+                    }
+                }
+            }
```
Comment (vados-cosmonic) [inline crates/provider-wit-bindgen-macro/src/bindgen_visitor.rs:433]: @connorsmith256 this is the most important bit -- the structure of generated code changed, and where the methods show up and how they needed to be processed had to be adjusted. This code was previously dealing primarily with`Item::Struct`s

File: crates/provider-wit-bindgen-macro/src/rust.rs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+            tokens.append_all([
+                &arr_ref[0], // name
+                &arr_ref[1], // colon
+                &TokenTree::Ident(Ident::new("Vec", Span::call_site())), // Vec
+                &TokenTree::Punct(Punct::new('<', Spacing::Joint)), // <
+            ]);
+            tokens.extend(t.stream());
+            tokens.append(TokenTree::Punct(Punct::new('>', Spacing::Joint))); // >
+        },
```
Comment (vados-cosmonic) [inline crates/provider-wit-bindgen-macro/src/rust.rs:105]: @connorsmith256 this case was added -- `&[String]`, essentially

File: crates/provider-wit-bindgen-macro/src/lib.rs [hunk 3]
```diff
@@ -1494,925 +535,45 @@ pub fn generate(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
-                        panic!("found duplicate instances of enum [${}]", e.ident);
-                    }
-
-                    self.serde_extended_enums
-                        .insert(e.ident.to_string(), (import_path, e.clone()));
-                }
-            }
-
-            // Process struct declarations that appear in the bindgen output
-            Item::Struct(s) => {
```
Comment (vados-cosmonic) [inline crates/provider-wit-bindgen-macro/src/lib.rs:2034]: @connorsmith256 here's that old `Item::Struct` code

**Comment by vados-cosmonic** [comment]
Sure @connorsmith256 -- noted some areas that saw change -- most of the rest was splitting up the functionality

**Pull request merged**